crates/ruff/src/rules/ruff/rules/ambiguous_unicode_character.rs

```diff
@@ -103,50 +101,133 @@ pub(crate) fn ambiguous_unicode_character(
 
     let text = locator.slice(range);
 
+    // Most of the time, we don't need to check for ambiguous unicode characters at all.
+    if text.is_ascii() {
+        return diagnostics;
+    }
+
+    // Iterate over the "words" in the text.
+    let mut flags = WordFlags::empty();
+    let mut buffer = vec![];
     for (relative_offset, current_char) in text.char_indices() {
-        if !current_char.is_ascii() {
-            // Search for confusing characters.
-            if let Some(representant) = CONFUSABLES.get(&(current_char as u32)).copied() {
-                if !settings.allowed_confusables.contains(&current_char) {
-                    let char_range = TextRange::at(
+        // Word boundary.
+        if !current_char.is_alphanumeric() {
+            if !buffer.is_empty() {
+                if flags.is_candidate_word() {
+                    diagnostics.append(&mut buffer);
+                }
+                buffer.clear();
+            }
```

Comment:
Nit: This could be slightly faster if we only store the `TextRange`, confusable, and represent in the `buffer` and delay the generation of the diagnostics until here.

Reply:
Would that mean we would have to pass the confusable and represent to this function, or would there be any other way to add it to the buffer. (just trying to understand the codebase a bit more)

Reply:
The idea would be to change `buffer` to `Vec<Candidate>` where `Candidate` is defined as 

```rust
struct Candidate {
	range: TextRange,
	confusable: char,
	representant: char
}
```

And change this loop here to:

```rust
if !word_candidates.is_empty() {
	if flags.is_candidate_word() {
		for candidate in word_candidates.drain() {
			let kind = match context { ... };
			if settings.enabled(kind) {
				diagnostics.push(Diagnostic {  ... })
			}
		}
	}
	
	word_candidates.clear();
}

```

Reply:
Wouldn't a `Vec<usize>` for `relative_offset` alone be enough? Once we know we know we want to flag the current word and actually need to emit diagnostics we can read confusable and representant from the word

Reply:
Storing the `relative_offset` requires that we then do the lookup again, I think? I opted for `Candidate` for now.